# Mandatory Business Case Study 2 By: Vidhyadhar Melam

## Problem: Bank Marketing Business Case Analysis
This dataset captures the results of a series of direct marketing campaigns of “BANK”, an
international banking institution. The marketing campaigns were based on phone calls. Often,
more than one contact to the same client was required, in order to access if the product (bank
term deposit) would be ('yes') or not ('no') subscribed.
The data is related with direct marketing campaigns of a Portuguese banking institution. The
marketing campaigns were based on phone calls. Often, more than one contact to the same
client was required, in order to access if the product (bank term deposit) would be ('yes') or
not ('no') subscribed.

### Questions: Solve Using logistic regression method:
   
1. Run 2 models that predict the subscription of a term deposit
2. Create an appropriate split and validate your models using ratio of correct predictions vs total predictions.
3. Asses your model using performance metrics such as accuracy, precision, recall,F-score and G-score.
4. Choose one model and provide actionable recommendations to the marketing department.

#1. Run 2 models that predict the subscription of a term deposit
    SOLUTION :

In [19]:
DF = read.csv("C:\\Users\\viddi\\Desktop\\BUSINESS CASE STUDY 2\\Datasets\\Assignment_2.csv",na.strings = "")

In [20]:
str(DF)

'data.frame':	41188 obs. of  21 variables:
 $ age           : int  56 57 37 40 56 45 59 41 24 25 ...
 $ job           : Factor w/ 12 levels "admin.","blue-collar",..: 4 8 8 1 8 8 1 2 10 8 ...
 $ marital       : Factor w/ 4 levels "divorced","married",..: 2 2 2 2 2 2 2 2 3 3 ...
 $ education     : Factor w/ 8 levels "basic.4y","basic.6y",..: 1 4 4 2 4 3 6 8 6 4 ...
 $ default       : Factor w/ 3 levels "no","unknown",..: 1 2 1 1 1 2 1 2 1 1 ...
 $ housing       : Factor w/ 3 levels "no","unknown",..: 1 1 3 1 1 1 1 1 3 3 ...
 $ loan          : Factor w/ 3 levels "no","unknown",..: 1 1 1 1 3 1 1 1 1 1 ...
 $ contact       : Factor w/ 2 levels "cellular","telephone": 2 2 2 2 2 2 2 2 2 2 ...
 $ month         : Factor w/ 10 levels "apr","aug","dec",..: 7 7 7 7 7 7 7 7 7 7 ...
 $ day_of_week   : Factor w/ 5 levels "fri","mon","thu",..: 2 2 2 2 2 2 2 2 2 2 ...
 $ duration      : int  261 149 226 151 307 198 139 217 380 50 ...
 $ campaign      : int  1 1 1 1 1 1 1 1 1 1 ...
 $ pdays         : i

In [21]:
model_1 = glm(y~ age+job+marital+education+default+housing+loan+contact+
             month+day_of_week+campaign+pdays+previous+poutcome+
             emp.var.rate+cons.price.idx+cons.conf.idx+euribor3m+nr.employed,data = DF,family="binomial")
model_2 = glm(y~ age+job+marital+education+default+housing+loan+contact+
             month+day_of_week+campaign+pdays+previous+poutcome+
             emp.var.rate+cons.price.idx+cons.conf.idx+euribor3m+nr.employed,data = DF,family="quasibinomial")

In [22]:
summary(model_1)


Call:
glm(formula = y ~ age + job + marital + education + default + 
    housing + loan + contact + month + day_of_week + campaign + 
    pdays + previous + poutcome + emp.var.rate + cons.price.idx + 
    cons.conf.idx + euribor3m + nr.employed, family = "binomial", 
    data = DF)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.0810  -0.3926  -0.3190  -0.2617   2.9643  

Coefficients: (1 not defined because of singularities)
                               Estimate Std. Error z value Pr(>|z|)    
(Intercept)                  -2.285e+02  3.339e+01  -6.844 7.68e-12 ***
age                           4.392e-04  2.128e-03   0.206 0.836459    
jobblue-collar               -1.515e-01  6.882e-02  -2.201 0.027704 *  
jobentrepreneur              -7.367e-02  1.074e-01  -0.686 0.492598    
jobhousemaid                 -9.238e-02  1.285e-01  -0.719 0.472111    
jobmanagement                -6.176e-02  7.520e-02  -0.821 0.411493    
jobretired                    2.369e-01  9

In [23]:
summary(model_2)


Call:
glm(formula = y ~ age + job + marital + education + default + 
    housing + loan + contact + month + day_of_week + campaign + 
    pdays + previous + poutcome + emp.var.rate + cons.price.idx + 
    cons.conf.idx + euribor3m + nr.employed, family = "quasibinomial", 
    data = DF)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.0810  -0.3926  -0.3190  -0.2617   2.9643  

Coefficients: (1 not defined because of singularities)
                               Estimate Std. Error t value Pr(>|t|)    
(Intercept)                  -2.285e+02  3.361e+01  -6.798 1.07e-11 ***
age                           4.392e-04  2.142e-03   0.205 0.837545    
jobblue-collar               -1.515e-01  6.928e-02  -2.187 0.028774 *  
jobentrepreneur              -7.367e-02  1.081e-01  -0.682 0.495515    
jobhousemaid                 -9.238e-02  1.293e-01  -0.714 0.475099    
jobmanagement                -6.176e-02  7.571e-02  -0.816 0.414650    
jobretired                    2.369e-

#2. Create an appropriate split and validate your models using ratio of correct predictions vs total predictions.


In [24]:
library(dplyr)

In [25]:
## 75% of the sample size
smp_size <- floor(0.75 * nrow(DF))

## set the seed to make your partition reproducible
set.seed(123)
train_ind <- sample(seq_len(nrow(DF)), size = smp_size)

Train <- DF[train_ind, ]
Test <- DF[-train_ind, ]

In [26]:
model_1 = glm(y~ age+job+marital+education+default+housing+loan+contact+
             month+day_of_week+campaign+pdays+previous+poutcome+
             emp.var.rate+cons.price.idx+cons.conf.idx+euribor3m+nr.employed,data = Train,family="binomial")
model_2 = glm(y~ age+job+marital+education+default+housing+loan+contact+
             month+day_of_week+campaign+pdays+previous+poutcome+
             emp.var.rate+cons.price.idx+cons.conf.idx+euribor3m+nr.employed,data = Train,family="quasibinomial")

In [27]:
Test$model_prob <- predict(model_1, type = "response",newdata = Test)


Warning message in predict.lm(object, newdata, se.fit, scale = 1, type = if (type == :
"prediction from a rank-deficient fit may be misleading"

In [28]:
Test_1 <- Test  %>% mutate(model_pred = 1*(model_prob > .50) + 0, y_binary = 1*(y == "yes") + 0)

In [29]:
Test_1 <- Test_1 %>% mutate(accurate = 1*(model_pred == y_binary))
cat("Correct Prediction vs Total Prediction = ",sum(Test_1$accurate)/nrow(Test_1))


Correct Prediction vs Total Prediction =  0.9040497

> ### The above is test accuracy which is Correct prediction/ total predictions. Accuracy is 90.4%.

#3. Asses your model using performance metrics such as accuracy, precision, recall, F-score and G-score.

> #### From the above solution we know that accuracy of the model is 90%

In [30]:
P = nrow(Test_1[Test_1$y_binary == 1,])
N = nrow(Test_1[Test_1$y_binary == 0,])
TP = nrow(Test_1[Test_1$y_binary == 1& Test_1$model_pred==1,])
TN = nrow(Test_1[Test_1$y_binary == 0& Test_1$model_pred==0,])
FN = nrow(Test_1[Test_1$y_binary == 1& Test_1$model_pred==0,])
FP = nrow(Test_1[Test_1$y_binary == 0& Test_1$model_pred==1,])

In [31]:
Sensitivity.TPR = TP / (TP + FN)
Specificity.SPC = TN / (FP + TN)
Precision.PPV = TP / (TP + FP)
Negative_Predictive_Value.NPV = TN / (TN + FN)
False_Positive_Rate.FPR = FP / (FP + TN)
False_Discovery_Rate.FDR = FP / (FP + TP)
False_Negative_Rate.FNR = FN / (FN + TP)
Accuracy.ACC = (TP + TN) / (P + N)
F1_Score.F1 = 2*TP / (2*TP + FP + FN)
recall= TP/(TP+FN)

In [32]:
cat("Accuracy of the model is = ",Accuracy.ACC*100)

Accuracy of the model is =  90.40497

In [33]:
cat("Recall of the model is = ", recall*100)

Recall of the model is =  23.29137

In [34]:
cat("The F score for the model is =", F1_Score.F1*100)

The F score for the model is = 34.39575

In [35]:
Sensitivity.TPR

[1] 0.2329137

In [36]:
Precision.PPV

[1] 0.6573604

In [37]:
Specificity.SPC

[1] 0.9853021

In [38]:
G.score = sqrt(Precision.PPV*recall)
cat("The G score for the model is = ",G.score*100)

The G score for the model is =  39.12905